OpenRA.Mods.Common/Widgets/ObserverArmyIconsWidget.cs

```diff
@@ -0,0 +1,271 @@
+ * This file is part of OpenRA, which is free software. It is made
+ * available to you under the terms of the GNU General Public License
+ * as published by the Free Software Foundation, either version 3 of
+ * the License, or (at your option) any later version. For more
+ * information, see COPYING.
+ */
+#endregion
+
+using System;
+using System.Collections.Generic;
+using System.Linq;
+using OpenRA.Graphics;
+using OpenRA.Mods.Common.Traits;
+using OpenRA.Mods.Common.Traits.Render;
+using OpenRA.Primitives;
+using OpenRA.Widgets;
+
+namespace OpenRA.Mods.Common.Widgets
+{
+	public class ObserverArmyIconsWidget : Widget
+	{
+		public Func<Player> GetPlayer;
+		readonly World world;
+		readonly WorldRenderer worldRenderer;
+
+		public int IconWidth = 32;
+		public int IconHeight = 24;
+		public int IconSpacing = 1;
+
+		Dictionary<string, Tuple<ActorInfo, int, int>> unitLookup;
+		float2 iconSize;
+		public int MinWidth = 240;
+
+		public ActorInfo TooltipUnit { get; private set; }
+		public Func<ActorInfo> GetTooltipUnit;
+
+		public readonly string TooltipTemplate = "ARMY_TOOLTIP";
+		public readonly string TooltipContainer;
+
+		readonly Lazy<TooltipContainerWidget> tooltipContainer;
+		readonly List<ArmyIcon> armyIcons = new List<ArmyIcon>();
+
+		readonly CachedTransform<Player, PlayerStatistics> stats = new CachedTransform<Player, PlayerStatistics>(player => player.PlayerActor.TraitOrDefault<PlayerStatistics>());
+		readonly Cache<string, ActorData> actorDataCache;
+
+		int lastIconIdx;
+		Rectangle currentIconBounds;
+
+		[ObjectCreator.UseCtor]
+		public ObserverArmyIconsWidget(World world, WorldRenderer worldRenderer)
+		{
+			this.world = world;
+			this.worldRenderer = worldRenderer;
+
+			GetTooltipUnit = () => TooltipUnit;
+			tooltipContainer = Exts.Lazy(() =>
+				Ui.Root.Get<TooltipContainerWidget>(TooltipContainer));
+
+			Init();
+
+			actorDataCache = new Cache<string, ActorData>(LoadActorDataCache);
+		}
+
+		ActorData LoadActorDataCache(string name)
+		{
+			var actorInfo = unitLookup[name].Item1;
+			var bi = actorInfo.TraitInfo<BuildableInfo>();
+
+			var rsi = actorInfo.TraitInfo<RenderSpritesInfo>();
+
+			return new ActorData(bi, rsi);
+		}
+
+		protected ObserverArmyIconsWidget(ObserverArmyIconsWidget other)
+			: base(other)
+		{
+			GetPlayer = other.GetPlayer;
+			world = other.world;
+			worldRenderer = other.worldRenderer;
+
+			IconWidth = other.IconWidth;
+			IconHeight = other.IconHeight;
+			IconSpacing = other.IconSpacing;
+			iconSize = new float2(IconWidth, IconHeight);
+
+			MinWidth = other.MinWidth;
+
+			TooltipUnit = other.TooltipUnit;
+			GetTooltipUnit = () => TooltipUnit;
+
+			TooltipTemplate = other.TooltipTemplate;
+			TooltipContainer = other.TooltipContainer;
+
+			tooltipContainer = Exts.Lazy(() =>
+				Ui.Root.Get<TooltipContainerWidget>(TooltipContainer));
+
+			Init();
+
+			actorDataCache = new Cache<string, ActorData>(LoadActorDataCache);
+		}
+
+		void Init()
```

Comment:
This is a bit of a monster. What do you think about 826a91ef83e31e71962ab194110f6f5753f71740 to simplify this?